src/app/core/services/notifications/app-server-rest-notification.service.ts

```diff
@@ -0,0 +1,135 @@
+      this.config.getSourceID()
+    ]).then(([user, sourceId]) => Object.assign({}, user, { sourceId }))
+  }
+
+  publishAllNotifications(user, limit): Promise<any> {
+    return this.schedule.getTasks(TaskType.ALL).then(tasks => {
+      const fcmNotifications = this.notifications
+        .futureNotifications(tasks, limit)
+        .map(t => this.format(t, user))
+      this.logger.log('NOTIFICATIONS Scheduling FCM notifications')
+      this.logger.log(fcmNotifications)
+      return Promise.all(
+        fcmNotifications
+          .map(n => this.sendNotification(n, user.subjectId, user.projectId))
+          .concat([this.setLastNotificationUpdate(Date.now())])
+      )
+    })
+  }
+
+  publishTestNotification(user): Promise<void> {
+    return this.sendNotification(
+      this.format(this.notifications.createTestNotification(), user),
+      user.subjectId,
+      user.projectId
+    )
+  }
+
+  cancelAllNotifications(user): Promise<any> {
+    return this.appServerService.getApiClient().then(apiClient =>
+      apiClient.apis[
+        this.FCM_NOTIFICATION_CONTROLLER
+      ].deleteNotificationsForUser({
+        subjectId: user.subjectId,
+        projectId: user.projectId
+      })
+    )
+  }
+
+  cancelSingleNotification(user, notificationId) {
+    return this.appServerService.getApiClient().then(apiClient =>
+      apiClient.apis[this.FCM_NOTIFICATION_CONTROLLER]
+        .deleteNotificationUsingProjectIdAndSubjectIdAndNotificationId({
+          subjectId: user.subjectId,
+          projectId: user.projectId,
+          id: notificationId
+        })
+        .then(() =>
+          console.log('Success cancelling notification ' + notificationId)
+        )
+    )
+  }
+
+  private sendNotification(notification, subjectId, projectId): Promise<any> {
+    return this.appServerService.getApiClient().then(apiClient =>
+      apiClient.apis[this.FCM_NOTIFICATION_CONTROLLER]
+        .addSingleNotification(
+          { projectId: projectId, subjectId: subjectId },
+          { requestBody: notification.notificationDto }
+        )
+        .then(res => {
+          notification.notification.id = res.body.id
+          return this.logger.log('Successfully sent! Updating notification Id')
+        })
+        .catch(err => {
+          this.logger.error('Failed to send notification', err)
+          if (this.upstreamResends++ < DefaultMaxUpstreamResends)
+            this.sendNotification(notification, subjectId, projectId)
+        })
+    )
+  }
+
+  private format(notification: SingleNotification, user) {
+    const taskInfo = notification.task
+    return {
+      notification,
+      notificationDto: {
+        title: notification.title,
+        body: notification.text,
+        ttlSeconds: this.calculateTtlSeconds(
+          taskInfo.timestamp,
+          notification.timestamp,
+          taskInfo.completionWindow
+        ),
+        sourceId: user.sourceId,
+        type: taskInfo.name,
```

Comment:
why not call it a task-type instead of type? At least for me, at a glance what I thought as types would be Notification or Message.

Reply:
This is from the app server notification object. I think this is so that it's more generalized for future use cases when the notification isn't necessarily linked to a task? But I'm not sure. What do you think @yatharthranjan?